src/mlpack/methods/lsh_model/lshmodel.hpp

```diff
@@ -0,0 +1,373 @@
+   * @param A perturbation set to expand.
+  */
+  bool PerturbationExpand(std::vector<bool>& A) const;
+
+  /**
+   * Return true if perturbation set A is valid. A perturbation set is invalid if
+   * it contains two (or more) actions for the same dimension or dimensions that
+   * are larger than the queryCode's dimensions.
+   * @param A perturbation set to validate.
+   * @param numProj The number of projections for the sequence under validation.
+  */
+  bool PerturbationValid(const std::vector<bool>& A, size_t numProj) const;
+   /**
+    * Function that creates a template perturbation sequence given a value for
+    * an M and a W. The template perturbation sequence is based on the
+    * statistical properties of multi-probe LSH and uses those, instead of
+    * specific points, to generate scores.
+    * See mlpack/methods/lsh/lsh_search_impl.hpp for more details about how
+    * perturbation sequences are generated from specific points.
+    *
+    * @param numProj The number of projections for the LSH scheme for which we
+    *     want to compute the template perturbation sequence.
+    * @param hashWidth The hash width for the LSH scheme.
+    * @param numProbes The number of probes to generate.
+    */
+   void GenerateTemplateSequence(size_t numProj, 
+                                 double hashWidth, 
+                                 size_t numProbes);
+
+   /** Matrix that stores, in each column, the "direction" of the perturbation:
+    * 0 means no perturbation on that dimension, -1 means reduce dimension value
+    * by 1, and +1 means increase dimension value by 1.
+    */
+   
+   arma::Mat<short int> templateSequence;
+
+   /**
+    * Function that fits two DistanceStatisticPredictors - one
+    * to predict arithmetic mean and one to preduct geometric mean.
+    *
+    * @param referenceSizes The number of reference points for each kNN search.
+    * @param kValues The rank of the neighbors used for the statistic, for
+    *     example k = 5 means Ek is the arithmetic mean of the 5th-nearest
+    *     neighbor for different sample sizes.
+    * @param Ek The arithmetic mean of the squared distances of a point and its
+    *      k-nearest neighbor. One column per k.
+    * @param Gk The geometric mean of the squared distances of a point and its
+    *      k-nearest neighbor. One column per k.
+    */
+   void ApproximateKNNStatistics(const arma::Col<size_t>& referenceSizes, 
+                                 const arma::Col<size_t>& kValues,
+                                 const arma::mat& Ek, 
+                                 const arma::mat& Gk);
+
+   /**
+    * This is a helper class that uses the function a * k^b * N^c for some
+    * parameters a, b, c that have been fit to either predict the arithmetic or
+    * geometric mean of the squared distance of a point to its k-nearest
+    * neighbor, given some dataset size N and its k-nearest neighbor.
+    */
+   class DistanceStatisticPredictor
```

Comment:
If you like, you can make this file smaller by moving `DistanceStatisticPredictor` into a different file and making it not a nested class of `LSHModel`.  I'm ambivalent either way.